blockdag/reachability.go

```diff
@@ -547,9 +811,276 @@ func (dag *BlockDAG) updateReachability(node *blockNode, selectedParentAnticone
 			return err
 		}
 	}
+
+	// Update the reindex root.
+	// Note that we check for blue score here in order to find out
+	// whether the new node is going to be the virtual's selected
+	// parent. We don't check node == virtual.selectedParent because
+	// at this stage the virtual had not yet been updated.
+	if node.blueScore > dag.SelectedTipBlueScore() {
+		updateStartTime := time.Now()
+		modifiedTreeNodes, err := dag.updateReachabilityReindexRoot(newTreeNode)
+		if err != nil {
+			return err
+		}
+		if len(modifiedTreeNodes) > 0 {
+			updateTimeElapsed := time.Since(updateStartTime)
+			log.Debugf("Reachability reindex root updated to %s. "+
+				"Modified %d tree nodes and took %dms.",
+				dag.reachabilityReindexRoot.blockNode.hash,
+				len(modifiedTreeNodes), updateTimeElapsed.Milliseconds())
+			for modifiedTreeNode := range modifiedTreeNodes {
+				dag.reachabilityStore.setTreeNode(modifiedTreeNode)
+			}
+		}
+	}
+
 	return nil
 }
 
+func (dag *BlockDAG) updateReachabilityReindexRoot(newTreeNode *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	nextReindexRoot := dag.reachabilityReindexRoot
+	for {
+		candidateReindexRoot, modifiedNodes, found, err := dag.tryMovingReachabilityReindexRoot(nextReindexRoot, newTreeNode)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			break
+		}
+		for node := range modifiedNodes {
+			modifiedTreeNodes[node] = struct{}{}
+		}
+		nextReindexRoot = candidateReindexRoot
+	}
+
+	dag.reachabilityReindexRoot = nextReindexRoot
+	return modifiedTreeNodes, nil
+}
+
+func (dag *BlockDAG) tryMovingReachabilityReindexRoot(
+	reindexRoot *reachabilityTreeNode, newTreeNode *reachabilityTreeNode) (
+	newReindexRoot *reachabilityTreeNode, modifiedTreeNodes map[*reachabilityTreeNode]struct{}, found bool, err error) {
+
+	if !reindexRoot.isAncestorOf(newTreeNode) {
+		commonAncestor := reindexRoot.findCommonAncestor(newTreeNode)
+		return commonAncestor, nil, true, nil
+	}
+
+	chosenReindexRootChild, err := reindexRoot.findReachabilityTreeAncestorInChildren(newTreeNode)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	if newTreeNode.blockNode.blueScore-chosenReindexRootChild.blockNode.blueScore < reachabilityReindexWindow {
+		return nil, nil, false, nil
+	}
+	modifiedTreeNodes, err = dag.concentrateReachabilityTreeIntervalAroundReindexRootChild(reindexRoot, chosenReindexRootChild)
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	return chosenReindexRootChild, modifiedTreeNodes, true, nil
+}
+
+// findReachabilityTreeAncestorInChildren finds the reachability tree child
+// of rtn that is the ancestor of node.
+func (rtn *reachabilityTreeNode) findReachabilityTreeAncestorInChildren(node *reachabilityTreeNode) (*reachabilityTreeNode, error) {
```

Comment:
The whole point is to make the search in children logarithmic, so if this line copies the entire list into a futureCoveringSet it becomes linear anyway, so might as well just search linearly in the original children list... instead I suggest extracting findIndex logic into a global function which can be used here as well